casr/src/triage.rs

```diff
@@ -0,0 +1,464 @@
+            args[at_index + offset] = input;
+        }
+
+        let mut casr_cmd = Command::new(tool);
+        casr_cmd.args(&args);
+        casr_cmd.envs(envs);
+
+        debug!("{:?}", casr_cmd);
+
+        // Get output
+        let casr_output = casr_cmd
+            .output()
+            .with_context(|| format!("Couldn't launch {casr_cmd:?}"))?;
+
+        if !casr_output.status.success() {
+            let err = String::from_utf8_lossy(&casr_output.stderr);
+            if err.contains("Timeout") {
+                let mut timeout_name = self
+                    .path
+                    .file_name()
+                    .unwrap()
+                    .to_os_string()
+                    .into_string()
+                    .unwrap();
+                if let Some(idx) = timeout_name.find('-') {
+                    timeout_name.replace_range(..idx, "timeout");
+                }
+                let timeout_path = output_dir.join("timeout").join(timeout_name);
+                if fs::copy(&self.path, timeout_path).is_err() {
+                    error!("Error occurred while copying the file: {:?}", self.path);
+                }
+            } else if err.contains("Out of memory") {
+                let mut oom_name = self
+                    .path
+                    .file_name()
+                    .unwrap()
+                    .to_os_string()
+                    .into_string()
+                    .unwrap();
+                if let Some(idx) = oom_name.find('-') {
+                    oom_name.replace_range(..idx, "oom");
+                }
+                let oom_path = output_dir.join("oom").join(oom_name);
+                if fs::copy(&self.path, oom_path).is_err() {
+                    error!("Error occurred while copying the file: {:?}", self.path);
+                }
+            } else if err.contains("Program terminated (no crash)") {
+                warn!("{}: No crash on input {}", tool_name, self.path.display());
+            } else {
+                error!("{} for input: {}", err.trim(), self.path.display());
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// Perform crash analysis pipeline: Create, deduplicate and cluster CASR reports.
+///
+/// # Arguments
+///
+/// * `matches` - casr-afl/casr-libfuzzer arguments
+///
+/// * `crashes` - map of crashes, specified as a HashMap, where
+///               key is crash input file name and value is CrashInfo structure
+pub fn fuzzing_crash_triage_pipeline(
+    matches: &clap::ArgMatches,
+    crashes: &HashMap<String, CrashInfo>,
+) -> Result<()> {
+    // Get casr-cluster path
+    let casr_cluster = get_path("casr-cluster")?;
+
+    let mut envs = HashMap::new();
+    if crashes.is_empty() {
+        bail!("No crashes found");
+    }
+    let crash_info = crashes.values().next().unwrap();
+    if crash_info
+        .target_args
+        .iter()
+        .any(|x| x.contains("-detect_leaks=0"))
```

Comment:
We'd better check this stuff in run_casr function

Reply:
done